iree/compiler/Dialect/Flow/Transforms/OptimizeNumerics.cpp

```diff
@@ -0,0 +1,281 @@
+  using OpInterfaceRewritePattern::OpInterfaceRewritePattern;
+
+  LogicalResult matchAndRewrite(IREE::Util::NumericCastOpInterface castOp,
+                                PatternRewriter &rewriter) const override {
+    auto initTensorOp = castOp.getInput().getDefiningOp<linalg::InitTensorOp>();
+    if (!initTensorOp) return failure();
+    Type resultType = castOp.getCasted().getType();
+
+    rewriter.replaceOpWithNewOp<linalg::InitTensorOp>(
+        castOp, resultType, initTensorOp.sizes(), initTensorOp.static_sizes());
+    return success();
+  }
+};
+
+// For a cast produced by a fill, rewrites the cast to be on the fill operands.
+struct LinalgFillCast
+    : public OpInterfaceRewritePattern<IREE::Util::NumericCastOpInterface> {
+  using OpInterfaceRewritePattern::OpInterfaceRewritePattern;
+
+  LogicalResult matchAndRewrite(IREE::Util::NumericCastOpInterface castOp,
+                                PatternRewriter &rewriter) const override {
+    auto loc = castOp.getLoc();
+    auto fillOp = castOp.getInput().getDefiningOp<linalg::FillOp>();
+    if (!fillOp) return failure();
+    Type toElementType = getElementTypeOrSelf(castOp.getCastedType());
+
+    Value fillInput = fillOp.value();
+    Value fillInit = fillOp.output();
+    fillInput = castOp
+                    .cloneWithInput(
+                        rewriter,
+                        withNewElementType(fillInput.getType(), toElementType),
+                        fillInput)
+                    .getCasted();
+    fillInit =
+        castOp
+            .cloneWithInput(
+                rewriter, withNewElementType(fillInit.getType(), toElementType),
+                fillInit)
+            .getCasted();
+    Value fillResult =
+        rewriter.create<linalg::FillOp>(loc, fillInput, fillInit).result();
+    rewriter.replaceOp(castOp, fillResult);
+    return success();
+  }
+};
+
+// For narrowable inputs, selects
+struct LinalgFpMatmulToLowP : public OpRewritePattern<linalg::MatmulOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = matmulOp.getLoc();
+    Type origResultType = matmulOp.getResult(0).getType();
+    auto lhsParams = NarrowParams::forValue(matmulOp.inputs()[0]);
+    auto rhsParams = NarrowParams::forValue(matmulOp.inputs()[1]);
+    auto accumParams = NarrowParams::forValue(matmulOp.outputs()[0]);
+    if (!lhsParams || !rhsParams || !accumParams) {
+      return rewriter.notifyMatchFailure(matmulOp, "no narrowing annotations");
+    }
+
+    // TODO: This could be more flexible, allowing mix and match integer/float
+    // types.
+    if (!lhsParams->isFromFloat() || !rhsParams->isFromFloat()) {
+      return rewriter.notifyMatchFailure(matmulOp, "not from floating point");
+    }
+
+    // TODO: Could support partial conversion to integer.
```

Comment:
lots of TODO's in here - always good to either make these `TODO(laurenzo)` (note-to-self) or `TODO(#issue)` that can be tracked - naked TODOs never get done, while attributed ones at least have a fighting chance :)

Reply:
Well, each of these are potential optimizations that are quite well know/fought over, so I suspect someone will want them based on perf numbers at some point. Was just erring on the side of simple/correct. Filed https://github.com/google/iree/issues/7987 and annotated.